Add a task page for troubleshooting kubectl (#42347)
kubernetes/website: content/en/docs/tasks/debug/debug-cluster/troubleshoot-kubectl.md

---
reviewers:
# Should be added by an experienced contributor
title: "Troubleshooting kubectl"
content_type: task
weight: 10
---

<!-- overview -->

This documentation is about investigating and diagnosing
{{<glossary_tooltip text="kubectl" term_id="kubectl">}} related issues.
If you are unable to access `kubectl` or unable to connect to your cluster using
`kubectl`, this document describes several common scenarios and potential solutions

> to identify and address the likely cause.

Reviewer: Please consider refactor the above sentence ...
"If you are unable to do something, (then) this document describes ...",
the logic here reads confusing to me.

<!-- body -->

## {{% heading "prerequisites" %}}

* You need to have a Kubernetes cluster.
* You also need to have `kubcectl` installed.

## Verify kubectl setup

Make sure you have installed and configured `kubectl` correctly on your local machine.
Check the `kubectl` version to ensure it is up-to-date and compatible with your cluster.

Check kubectl version:

```shell
kubectl version
```

You'll see a similar output:

```console
Client Version: version.Info{Major:"1", Minor:"27", GitVersion:"v1.27.4",GitCommit:"fa3d7990104d7c1f16943a67f11b154b71f6a132", GitTreeState:"clean",BuildDate:"2023-07-19T12:20:54Z", GoVersion:"go1.20.6", Compiler:"gc", Platform:"linux/amd64"}
Kustomize Version: v5.0.1
Server Version: version.Info{Major:"1", Minor:"27", GitVersion:"v1.27.3",GitCommit:"25b4e43193bcda6c7328a6d147b1fb73a33f1598", GitTreeState:"clean",BuildDate:"2023-06-14T09:47:40Z", GoVersion:"go1.20.5", Compiler:"gc", Platform:"linux/amd64"}

```

If you see `Unable to connect to the server: dial tcp <server-ip>:8443: i/o timeout`,
instead of `Server Version`, you need to troubleshoot kubectl connectivity with your cluster.

> If you have installed the `kubectl` but your shell gives the `command not found` error

Reviewer: ```suggestion
If you have installed the `kubectl` but your shell gives the `command not found` error,
```

> make sure that the `$PATH` environment variable is correctly configured and you have

Reviewer: ```suggestion
make sure you have correctly configured the `$PATH` environment variable and you have
```

> installed the kubectl by following the [official documentation for installing kubectl](/docs/tasks/tools/#kubectl).

Reviewer: Make sure A and B ...
The implicit assumption is that A should happen before B, right?
How about swapping the A and B above?

If you are using a Virtual Private Network (VPN) to access your Kubernetes cluster,
make sure that your VPN connection is active and stable. Sometimes, VPN disconnections
can lead to connection issues with the cluster. Reconnect to the VPN and try accessing
the cluster again.

## Check kubeconfig

The `kubectl` requires a `kubeconfig` file to connect to a Kubernetes cluster. The
`kubeconfig` file is usually located under the `~/.kube/config` directory. Make sure
that you have a valid `kubeconfig` file. If you don't have a `kubeconfig` file, you can
create and configure it via `kubectl config` command. If you have deployed your Kubernetes
cluster on a cloud platform and lost your `kubeconfig` file, you can re-generate it using
your cloud provider's tools. Refer the cloud provider's documentation for re-generating
the `kubeconfig` file.

Check if the `$KUBECONFIG` environment variable is configured correctly. You can set
`$KUBECONFIG`environment variable or use the `--kubeconfig` parameter with the kubectl
to specify the directory of a `kubeconfig` file.

## Authentication and authorization

If the kubectl is returning an error regarding the authentication token or authentication

> server, validate the `$KUBERNETES_SERVER` and`$KUBERNETES_TOKEN` environment variables.

Reviewer: I'm not sure if token based authentication is widely used.
This is not always true, because an error related to authentication server doesn't
necessarily mean there are problems related to these environment variables.
When there is an assumption here, please spell it out explicitly.

Make sure that you are using the valid user credentials. And you have the permission to

> access the resource that you have requested.

Reviewer: The second sentence comes from no where. I haven't even try to access any resources.
What kind of an error will show up when I don't have permissions?
As a user, I need to tell the difference between authentication error and authorization error.

## Verify contexts

Kubernetes supports [multiple clusters and contexts](/docs/tasks/access-application-cluster/configure-access-multiple-clusters/).
Ensure that you are using the correct context to interact with your cluster.

List available contexts:

```shell
kubectl config get-contexts
```

Switch to the appropriate context:

```shell
kubectl config use-context <context-name>
```

## API server and load balancer

The {{<glossary_tooltip text="kube-apiserver" term_id="kube-apiserver">}} server is the

> central component of a Kubernetes cluster. If the API server or the load balancer it uses

Reviewer: ```suggestion
central component of a Kubernetes cluster. If the API server or the load balancer that
runs in front your API servers ...
```

is not reachable or not responding, you won't be able to interact with the cluster.

Check the if the API server's host is reachable by using `ping` command. Check cluster's

> network connectivity and firewall. Or if your are using a cloud provider for deploying

Reviewer: The previous two checks are about network connectivity issues, and the third one is about API server status. I'd not recommend using "Or" here...

the cluster, check your cloud provider's `health check status` for the API server.

Verify the status of the load balancer (if used) to ensure it is healthy and forwarding
traffic to the API server.

## TLS problems

> Transport Layer Security (TLS) is used to secure communication with the Kubernetes

Reviewer: "communication with the Kubernetes API server" is ambiguous ...
are you talking about the connection between my local machine and the API server,
or the connection that is forwarded by the LB?
Maybe we should instead saying that **by default**, the API server only serves HTTPS requests.

API server. TLS problems can occur due to various reasons, such as certificate expiry or
chain of trust validity.

You can find the TLS certificate in the kubeconfig file, located in the `~/.kube/config`
directory. The `certificate-authority` attribute contains the CA certificate and the
`client-certificate` attribute contains the client certificate.

> Verify the expiry of these certificates:

Reviewer: This verification is incomplete. The certificates are in the kubeconfig file,
why I'm running `openssl` against something like `/path/to/....` in the following commands?

Author: Thanks for these suggestions.

I force pushed the code, can you please review it?

---

> reviewers:
> # Should be added by an experienced contributor

Reviewer 2: ```suggestion
```

title: "Troubleshooting kubectl"
content_type: task
weight: 10
---

<!-- overview -->

This documentation is about investigating and diagnosing
{{<glossary_tooltip text="kubectl" term_id="kubectl">}} related issues.
If you encounter issues accessing `kubectl` or connecting to your cluster, this
document outlines various common scenarios and potential solutions to help
identify and address the likely cause.

<!-- body -->

## {{% heading "prerequisites" %}}

* You need to have a Kubernetes cluster.

> * You also need to have `kubcectl` installed.

Reviewer 2: ```suggestion
* You also need to have `kubectl` installed - see [Install Tools](/docs/tasks/tools/#kubectl).
```

## Verify kubectl setup

Make sure you have installed and configured `kubectl` correctly on your local machine.
Check the `kubectl` version to ensure it is up-to-date and compatible with your cluster.

Check kubectl version:

```shell
kubectl version
```

You'll see a similar output:

```console
Client Version: version.Info{Major:"1", Minor:"27", GitVersion:"v1.27.4",GitCommit:"fa3d7990104d7c1f16943a67f11b154b71f6a132", GitTreeState:"clean",BuildDate:"2023-07-19T12:20:54Z", GoVersion:"go1.20.6", Compiler:"gc", Platform:"linux/amd64"}
Kustomize Version: v5.0.1
Server Version: version.Info{Major:"1", Minor:"27", GitVersion:"v1.27.3",GitCommit:"25b4e43193bcda6c7328a6d147b1fb73a33f1598", GitTreeState:"clean",BuildDate:"2023-06-14T09:47:40Z", GoVersion:"go1.20.5", Compiler:"gc", Platform:"linux/amd64"}

```

If you see `Unable to connect to the server: dial tcp <server-ip>:8443: i/o timeout`,
instead of `Server Version`, you need to troubleshoot kubectl connectivity with your cluster.

Make sure you have installed the kubectl by following the
[official documentation for installing kubectl](/docs/tasks/tools/#kubectl), and you have

> properly configured the `$PATH` environment variable.

Reviewer 2: I would mention the need for a kubeconfig here, _before_ mentioning VPN issues.

## Check VPN connectivity

If you are using a Virtual Private Network (VPN) to access your Kubernetes cluster,
make sure that your VPN connection is active and stable. Sometimes, VPN disconnections
can lead to connection issues with the cluster. Reconnect to the VPN and try accessing
the cluster again.

## Check kubeconfig

The `kubectl` requires a `kubeconfig` file to connect to a Kubernetes cluster. The
`kubeconfig` file is usually located under the `~/.kube/config` directory. Make sure
that you have a valid `kubeconfig` file. If you don't have a `kubeconfig` file, you can
create and configure it via `kubectl config` command. If you have deployed your Kubernetes
cluster on a cloud platform and lost your `kubeconfig` file, you can re-generate it using
your cloud provider's tools. Refer the cloud provider's documentation for re-generating
the `kubeconfig` file.

> 

Reviewer 2: Does running `kubectl config` produce a valid kubeconfig? If not, can we give the reader some more hints?

Also see https://github.com/kubernetes/website/issues/27934

Author: Thanks @sftim, I have force pushed it. Can you please review it?

kubernetes authentication token and the authentication server address are configured
properly.

If kubectl is returning an error regarding the authorization, make sure that you are
using the valid user credentials. And you have the permission to access the resource
that you have requested.

## Verify contexts

Kubernetes supports [multiple clusters and contexts](/docs/tasks/access-application-cluster/configure-access-multiple-clusters/).
Ensure that you are using the correct context to interact with your cluster.

List available contexts:

```shell
kubectl config get-contexts
```

Switch to the appropriate context:

```shell
kubectl config use-context <context-name>
```

## API server and load balancer

The {{<glossary_tooltip text="kube-apiserver" term_id="kube-apiserver">}} server is the
central component of a Kubernetes cluster. If the API server or the load balancer that
runs in front of your API servers is not reachable or not responding, you won't be able
to interact with the cluster.

Check the if the API server's host is reachable by using `ping` command. Check cluster's
network connectivity and firewall. If your are using a cloud provider for deploying

> the cluster, check your cloud provider's `health check status` for the API server.

Reviewer 2: ```suggestion
the cluster, check your cloud provider's health check status for the cluster's API server.
```

may occur due to various reasons, such as certificate expiry or chain of trust validity.

You can find the TLS certificate in the kubeconfig file, located in the `~/.kube/config`
directory. The `certificate-authority` attribute contains the CA certificate and the
`client-certificate` attribute contains the client certificate.

Verify the expiry of these certificates:

```shell
openssl x509 -noout -dates -in $(kubectl config view --minify --output 'jsonpath={.clusters[0].cluster.certificate-authority}')
```

output:
```console
notBefore=Sep  2 08:34:12 2023 GMT
notAfter=Aug 31 08:34:12 2033 GMT
```

```shell
openssl x509 -noout -dates -in $(kubectl config view --minify --output 'jsonpath={.users[0].user.client-certificate}')
```

output:
```console
notBefore=Sep  2 08:34:12 2023 GMT
notAfter=Sep  2 08:34:12 2026 GMT
```

## Verify kubectl helpers

Some kubectl authentication helpers provide easy access to Kubernetes clusters. If you
have used such helpers and are facing connectivity issues, ensure that the necessary
configurations are still present.

Check kubectl configuration for authentication details:

```shell
kubectl config view
```

> If you previously used a helper tool (e.g., kubectl-oidc-login), ensure that it is still

Reviewer 2: ```suggestion
If you previously used a helper tool (for example: `kubectl-oidc-login`), ensure that it is still
```
